File: pyomo/steiner_5_3d.py
# MINLP written by GAMS Convert at 05/15/24 12:00:08
#
# Equation counts
#     Total        E        G        L        N        X        C        B
#        28       10       18        0        0        0        0        0
#
# Variable counts
#                  x        b        i      s1s      s2s       sc       si
#     Total     cont   binary  integer     sos1     sos2    scont     sint
#        45       27       18        0        0        0        0        0
# FX      0
#
# Nonzero counts
#     Total    const       NL
#       138       75       63
#
# Reformulation has removed 1 variable and 1 equation

from pyomo.environ import *

model = m = ConcreteModel()

m.x1 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x2 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x3 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x4 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x5 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x6 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x7 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x8 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x9 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x10 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x11 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x12 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x13 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x14 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x15 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x16 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x17 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x18 = Var(within=Reals, bounds=(0,1.7320508075688772), initialize=0)
m.x19 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x20 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x21 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x22 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x23 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x24 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x25 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x26 = Var(within=Reals, bounds=(0,1), initialize=0)
m.x27 = Var(within=Reals, bounds=(0,1), initialize=0)
m.b28 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b29 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b30 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b31 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b32 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b33 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b34 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b35 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b36 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b37 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b38 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b39 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b40 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b41 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b42 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b43 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b44 = Var(within=Binary, bounds=(0,1), initialize=0)
m.b45 = Var(within=Binary, bounds=(0,1), initialize=0)

m.obj = Objective(sense=minimize, expr= m.x1 + m.x2 + m.x3 + m.x4 + m.x5 + m.x6
    + m.x7 + m.x8 + m.x9 + m.x10 + m.x11 + m.x12 + m.x13 + m.x14 + m.x15 +
    m.x16 + m.x17 + m.x18)

m.e1 = Constraint(expr= -sqrt((m.x19 - m.x22)**2 + (m.x20 - m.x23)**2 + (m.x21
    - m.x24)**2) + m.x1 - 1.7320508075688772 * m.b28 >= -1.7320508075688772)
m.e2 = Constraint(expr= -sqrt((m.x19 - m.x25)**2 + (m.x20 - m.x26)**2 + (m.x21
    - m.x27)**2) + m.x2 - 1.7320508075688772 * m.b29 >= -1.7320508075688772)
m.e3 = Constraint(expr= -sqrt((m.x22 - m.x25)**2 + (m.x23 - m.x26)**2 + (m.x24
    - m.x27)**2) + m.x3 - 1.7320508075688772 * m.b30 >= -1.7320508075688772)
m.e4 = Constraint(expr= -sqrt((-0.965242141552123 + m.x19)**2 + (
    -0.011654693792141124 + m.x20)**2 + (-0.7359916197968754 + m.x21)**2) +
    m.x4 - 1.7320508075688772 * m.b31 >= -1.7320508075688772)
m.e5 = Constraint(expr= -sqrt((-0.15801272476474815 + m.x19)**2 + (
    -0.9863394516628233 + m.x20)**2 + (-0.016880654207976242 + m.x21)**2) +
    m.x5 - 1.7320508075688772 * m.b32 >= -1.7320508075688772)
m.e6 = Constraint(expr= -sqrt((-0.8794912681346712 + m.x19)**2 + (
    -0.6813506644014146 + m.x20)**2 + (-0.8573423824815248 + m.x21)**2) + m.x6
    - 1.7320508075688772 * m.b33 >= -1.7320508075688772)
m.e7 = Constraint(expr= -sqrt((-0.9998162341661194 + m.x19)**2 + (
    -0.23971518216378862 + m.x20)**2 + (-0.33807775725669986 + m.x21)**2) +
    m.x7 - 1.7320508075688772 * m.b34 >= -1.7320508075688772)
m.e8 = Constraint(expr= -sqrt((-0.7081943329851277 + m.x19)**2 + (
    -0.28058168584313725 + m.x20)**2 + (-0.2633142571474607 + m.x21)**2) + m.x8
    - 1.7320508075688772 * m.b35 >= -1.7320508075688772)
m.e9 = Constraint(expr= -sqrt((-0.965242141552123 + m.x22)**2 + (
    -0.011654693792141124 + m.x23)**2 + (-0.7359916197968754 + m.x24)**2) +
    m.x9 - 1.7320508075688772 * m.b36 >= -1.7320508075688772)
m.e10 = Constraint(expr= -sqrt((-0.15801272476474815 + m.x22)**2 + (
    -0.9863394516628233 + m.x23)**2 + (-0.016880654207976242 + m.x24)**2) +
    m.x10 - 1.7320508075688772 * m.b37 >= -1.7320508075688772)
m.e11 = Constraint(expr= -sqrt((-0.8794912681346712 + m.x22)**2 + (
    -0.6813506644014146 + m.x23)**2 + (-0.8573423824815248 + m.x24)**2) + m.x11
    - 1.7320508075688772 * m.b38 >= -1.7320508075688772)
m.e12 = Constraint(expr= -sqrt((-0.9998162341661194 + m.x22)**2 + (
    -0.23971518216378862 + m.x23)**2 + (-0.33807775725669986 + m.x24)**2) +
    m.x12 - 1.7320508075688772 * m.b39 >= -1.7320508075688772)
m.e13 = Constraint(expr= -sqrt((-0.7081943329851277 + m.x22)**2 + (
    -0.28058168584313725 + m.x23)**2 + (-0.2633142571474607 + m.x24)**2) +
    m.x13 - 1.7320508075688772 * m.b40 >= -1.7320508075688772)
m.e14 = Constraint(expr= -sqrt((-0.965242141552123 + m.x25)**2 + (
    -0.011654693792141124 + m.x26)**2 + (-0.7359916197968754 + m.x27)**2) +
    m.x14 - 1.7320508075688772 * m.b41 >= -1.7320508075688772)
m.e15 = Constraint(expr= -sqrt((-0.15801272476474815 + m.x25)**2 + (
    -0.9863394516628233 + m.x26)**2 + (-0.016880654207976242 + m.x27)**2) +
    m.x15 - 1.7320508075688772 * m.b42 >= -1.7320508075688772)
m.e16 = Constraint(expr= -sqrt((-0.8794912681346712 + m.x25)**2 + (
    -0.6813506644014146 + m.x26)**2 + (-0.8573423824815248 + m.x27)**2) + m.x16
    - 1.7320508075688772 * m.b43 >= -1.7320508075688772)
m.e17 = Constraint(expr= -sqrt((-0.9998162341661194 + m.x25)**2 + (
    -0.23971518216378862 + m.x26)**2 + (-0.33807775725669986 + m.x27)**2) +
    m.x17 - 1.7320508075688772 * m.b44 >= -1.7320508075688772)
m.e18 = Constraint(expr= -sqrt((-0.7081943329851277 + m.x25)**2 + (
    -0.28058168584313725 + m.x26)**2 + (-0.2633142571474607 + m.x27)**2) +
    m.x18 - 1.7320508075688772 * m.b45 >= -1.7320508075688772)
m.e19 = Constraint(expr= m.b31 + m.b36 + m.b41 == 1)
m.e20 = Constraint(expr= m.b32 + m.b37 + m.b42 == 1)
m.e21 = Constraint(expr= m.b33 + m.b38 + m.b43 == 1)
m.e22 = Constraint(expr= m.b34 + m.b39 + m.b44 == 1)
m.e23 = Constraint(expr= m.b35 + m.b40 + m.b45 == 1)
m.e24 = Constraint(expr= m.b28 + m.b29 + m.b31 + m.b32 + m.b33 + m.b34 + m.b35
    == 3)
m.e25 = Constraint(expr= m.b28 + m.b30 + m.b36 + m.b37 + m.b38 + m.b39 + m.b40
    == 3)
m.e26 = Constraint(expr= m.b29 + m.b30 + m.b41 + m.b42 + m.b43 + m.b44 + m.b45
    == 3)
m.e27 = Constraint(expr= m.b28 == 1)
m.e28 = Constraint(expr= m.b29 + m.b30 == 1)

File: jump/steiner_5_3d.jl
# MINLP written by GAMS Convert at 05/15/24 11:56:34
#
# Equation counts
#     Total        E        G        L        N        X        C        B
#        28       10       18        0        0        0        0        0
#
# Variable counts
#                  x        b        i      s1s      s2s       sc       si
#     Total     cont   binary  integer     sos1     sos2    scont     sint
#        45       27       18        0        0        0        0        0
# FX      0
#
# Nonzero counts
#     Total    const       NL
#       138       75       63
#
# Reformulation has removed 1 variable and 1 equation

using JuMP
using MathOptInterface
const MOI = MathOptInterface

model = m = Model()

@variable(m, 0 <= x1 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x2 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x3 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x4 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x5 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x6 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x7 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x8 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x9 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x10 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x11 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x12 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x13 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x14 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x15 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x16 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x17 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x18 <= 1.7320508075688772, start=0)
@variable(m, 0 <= x19 <= 1, start=0)
@variable(m, 0 <= x20 <= 1, start=0)
@variable(m, 0 <= x21 <= 1, start=0)
@variable(m, 0 <= x22 <= 1, start=0)
@variable(m, 0 <= x23 <= 1, start=0)
@variable(m, 0 <= x24 <= 1, start=0)
@variable(m, 0 <= x25 <= 1, start=0)
@variable(m, 0 <= x26 <= 1, start=0)
@variable(m, 0 <= x27 <= 1, start=0)
@variable(m, 0 <= b28 <= 1, binary=true, start=0)
@variable(m, 0 <= b29 <= 1, binary=true, start=0)
@variable(m, 0 <= b30 <= 1, binary=true, start=0)
@variable(m, 0 <= b31 <= 1, binary=true, start=0)
@variable(m, 0 <= b32 <= 1, binary=true, start=0)
@variable(m, 0 <= b33 <= 1, binary=true, start=0)
@variable(m, 0 <= b34 <= 1, binary=true, start=0)
@variable(m, 0 <= b35 <= 1, binary=true, start=0)
@variable(m, 0 <= b36 <= 1, binary=true, start=0)
@variable(m, 0 <= b37 <= 1, binary=true, start=0)
@variable(m, 0 <= b38 <= 1, binary=true, start=0)
@variable(m, 0 <= b39 <= 1, binary=true, start=0)
@variable(m, 0 <= b40 <= 1, binary=true, start=0)
@variable(m, 0 <= b41 <= 1, binary=true, start=0)
@variable(m, 0 <= b42 <= 1, binary=true, start=0)
@variable(m, 0 <= b43 <= 1, binary=true, start=0)
@variable(m, 0 <= b44 <= 1, binary=true, start=0)
@variable(m, 0 <= b45 <= 1, binary=true, start=0)

@objective(m, Min, x1 + x2 + x3 + x4 + x5 + x6 + x7 + x8 + x9 + x10 + x11 + x12
    + x13 + x14 + x15 + x16 + x17 + x18)

@NLconstraint(m, e1, -sqrt((x19 - x22)^2 + (x20 - x23)^2 + (x21 - x24)^2) + x1
    - 1.7320508075688772 * b28 >= -1.7320508075688772)
@NLconstraint(m, e2, -sqrt((x19 - x25)^2 + (x20 - x26)^2 + (x21 - x27)^2) + x2
    - 1.7320508075688772 * b29 >= -1.7320508075688772)
@NLconstraint(m, e3, -sqrt((x22 - x25)^2 + (x23 - x26)^2 + (x24 - x27)^2) + x3
    - 1.7320508075688772 * b30 >= -1.7320508075688772)
@NLconstraint(m, e4, -sqrt((-0.965242141552123 + x19)^2 + (
    -0.011654693792141124 + x20)^2 + (-0.7359916197968754 + x21)^2) + x4 -
    1.7320508075688772 * b31 >= -1.7320508075688772)
@NLconstraint(m, e5, -sqrt((-0.15801272476474815 + x19)^2 + (
    -0.9863394516628233 + x20)^2 + (-0.016880654207976242 + x21)^2) + x5 -
    1.7320508075688772 * b32 >= -1.7320508075688772)
@NLconstraint(m, e6, -sqrt((-0.8794912681346712 + x19)^2 + (-0.6813506644014146
    + x20)^2 + (-0.8573423824815248 + x21)^2) + x6 - 1.7320508075688772 * b33
    >= -1.7320508075688772)
@NLconstraint(m, e7, -sqrt((-0.9998162341661194 + x19)^2 + (
    -0.23971518216378862 + x20)^2 + (-0.33807775725669986 + x21)^2) + x7 -
    1.7320508075688772 * b34 >= -1.7320508075688772)
@NLconstraint(m, e8, -sqrt((-0.7081943329851277 + x19)^2 + (
    -0.28058168584313725 + x20)^2 + (-0.2633142571474607 + x21)^2) + x8 -
    1.7320508075688772 * b35 >= -1.7320508075688772)
@NLconstraint(m, e9, -sqrt((-0.965242141552123 + x22)^2 + (
    -0.011654693792141124 + x23)^2 + (-0.7359916197968754 + x24)^2) + x9 -
    1.7320508075688772 * b36 >= -1.7320508075688772)
@NLconstraint(m, e10, -sqrt((-0.15801272476474815 + x22)^2 + (
    -0.9863394516628233 + x23)^2 + (-0.016880654207976242 + x24)^2) + x10 -
    1.7320508075688772 * b37 >= -1.7320508075688772)
@NLconstraint(m, e11, -sqrt((-0.8794912681346712 + x22)^2 + (
    -0.6813506644014146 + x23)^2 + (-0.8573423824815248 + x24)^2) + x11 -
    1.7320508075688772 * b38 >= -1.7320508075688772)
@NLconstraint(m, e12, -sqrt((-0.9998162341661194 + x22)^2 + (
    -0.23971518216378862 + x23)^2 + (-0.33807775725669986 + x24)^2) + x12 -
    1.7320508075688772 * b39 >= -1.7320508075688772)
@NLconstraint(m, e13, -sqrt((-0.7081943329851277 + x22)^2 + (
    -0.28058168584313725 + x23)^2 + (-0.2633142571474607 + x24)^2) + x13 -
    1.7320508075688772 * b40 >= -1.7320508075688772)
@NLconstraint(m, e14, -sqrt((-0.965242141552123 + x25)^2 + (
    -0.011654693792141124 + x26)^2 + (-0.7359916197968754 + x27)^2) + x14 -
    1.7320508075688772 * b41 >= -1.7320508075688772)
@NLconstraint(m, e15, -sqrt((-0.15801272476474815 + x25)^2 + (
    -0.9863394516628233 + x26)^2 + (-0.016880654207976242 + x27)^2) + x15 -
    1.7320508075688772 * b42 >= -1.7320508075688772)
@NLconstraint(m, e16, -sqrt((-0.8794912681346712 + x25)^2 + (
    -0.6813506644014146 + x26)^2 + (-0.8573423824815248 + x27)^2) + x16 -
    1.7320508075688772 * b43 >= -1.7320508075688772)
@NLconstraint(m, e17, -sqrt((-0.9998162341661194 + x25)^2 + (
    -0.23971518216378862 + x26)^2 + (-0.33807775725669986 + x27)^2) + x17 -
    1.7320508075688772 * b44 >= -1.7320508075688772)
@NLconstraint(m, e18, -sqrt((-0.7081943329851277 + x25)^2 + (
    -0.28058168584313725 + x26)^2 + (-0.2633142571474607 + x27)^2) + x18 -
    1.7320508075688772 * b45 >= -1.7320508075688772)
@constraint(m, e19, b31 + b36 + b41 == 1)
@constraint(m, e20, b32 + b37 + b42 == 1)
@constraint(m, e21, b33 + b38 + b43 == 1)
@constraint(m, e22, b34 + b39 + b44 == 1)
@constraint(m, e23, b35 + b40 + b45 == 1)
@constraint(m, e24, b28 + b29 + b31 + b32 + b33 + b34 + b35 == 3)
@constraint(m, e25, b28 + b30 + b36 + b37 + b38 + b39 + b40 == 3)
@constraint(m, e26, b29 + b30 + b41 + b42 + b43 + b44 + b45 == 3)
@constraint(m, e27, b28 == 1)
@constraint(m, e28, b29 + b30 == 1)
